### PR TITLE
Refactor concepts APIs to quickly access optional instances and values properties

### PR DESCRIPTION
**Pull request by farost**
## Usage and product changes
We generalize the approach to getting concepts properties in TypeDB Drivers, introducing a set of new APIs for fetching optional values related to instances of `Concept` classes in Java and Python.

Now, all subclasses of `Concept` have a set of new interfaces starting with `try` to access IIDs, labels, value types, and values without a need to cast to a specific `Instance` or `Value`. These methods can be useful if:
* you have an established workflow with constant queries and always expect these properties to exist;
* you want to implement a custom handling of cases where the expected values are missing shorter (without exceptions).

Additionally, value type checks like `is_boolean`/`isBoolean` are also declared on the top-level `Concept`.

Note that casting is still possible, and its benefits are, as usual:
* static type checking for your programs using TypeDB Driver;
* access to the non-optional `get` interfaces for specific subclasses of `Concept` like `get_iid`/`getIID` for `Entity` and `Relation`.

Also note that some of the interfaces have changed (e.g., `asBoolean` -> `getBoolean` for `Attribute` and `Value` to separate value retrieval and concept casting), but their functioning has not.

## Implementation
On the C driver side, almost all subconcept-specific methods have been transformed into `concept_`, receiving just a `Concept` instance. The only exception is `get_type` for `Entity`, `Relation`, and `Attribute` instances, as it also uses the inner classes of the `Concept`'s enum in Rust (I think it's simple enough to leave it this way, but I may also refactor it if you want to).

On the wrapper's users' side (Python and Java), we use `try_get_` methods in `Concept` implementations and reuse these methods in `get_` methods in specific subclasses of `Concept` to throw or return the unwrapped error (sometimes we throw an inner error if it's our mission to make sure that the requested value exists, sometimes -- continue throwing user errors for unsafe gets like for `getBoolean`, just like we used to).

### CODE REVIEW

File: c/src/memory.rs [hunk 1]
```diff
@@ -47,6 +47,12 @@ pub(super) fn release_string(str: String) -> *mut c_char {
     raw
 }
 
+pub(super) fn release_optional_string(str: Option<String>) -> *mut c_char {
+    let raw = str.map(|str| CString::new(str).unwrap().into_raw()).unwrap_or_else(null_mut);
+    trace!("Releasing ownership of optional <CString> @ {:?}", raw);
+    raw
+}
```
Comment (dmitrii-ubskii) [inline c/src/memory.rs:54]:
```suggestion
pub(super) fn release_optional_string(str: Option<String>) -> *mut c_char {
    str.map(release_string).unwrap_or_else(null_mut)
}
```

**Review by dmitrii-ubskii — APPROVED** [review]
Looks good, left a couple comments / questions.

File: rust/src/answer/value_group.rs [hunk 2]
```diff
@@ -20,7 +20,7 @@
 use crate::concept::{Concept, Value};
 
 /// Contains an element of a TypeQL Get Group Aggregate query result.
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, PartialEq)] // TODO: Remove?
```
[the remark this answers is not in the document]
Comment (farost) [reply rust/src/answer/value_group.rs:23]: I've missed this TODO, thanks. The whole ValueGroup class, I guess, if it's not going to be used?

File: c/src/memory.rs (hunk 1 above)
Comment (farost) [reply c/src/memory.rs:54]: Will do

File: rust/src/answer/value_group.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmitrii-ubskii) [reply rust/src/answer/value_group.rs:23]: delet

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**